lib/spack/llnl/util/lang.py

```diff
@@ -227,48 +233,203 @@ def list_modules(directory, **kwargs):
+                    self.e
+                )
+
+            def __eq__(self, other):
+                return self._cmp_key() == other._cmp_key()
+
+            def __lt__(self):
+                return self._cmp_key() < other._cmp_key()
+
+            # etc.
+
+    Python would compare ``Widgets`` lexicographically based on their
+    tuples. The issue there for simple comparators is that we have to
+    bulid the tuples *and* we have to generate all the values in them up
+    front. When implementing comparisons for large data structures, this
+    can be costly.
+
+    Lazy lexicographic comparison maps the tuple comparison shown above
+    to generator functions. Instead of comparing based on pre-constructed
+    tuple keys, users of this decorator can compare using elements from a
+    generator. So, you'd write:
+
+        @lazy_lexicographic_ordering
+        class Widget:
+            def _cmp_iter(self):
+                yield a
+                yield b
+                def cd_fun():
+                    yield c
+                    yield d
+                yield cd_fun
+                yield e
+
+            # operators are added by decorator
+
+    There are no tuples preconstructed, and the generator does not have
+    to complete. Instead of tuples, we simply make functions that lazily
+    yield what would've been in the tuple. The
+    ``@lazy_lexicographic_ordering`` decorator handles the details of
+    implementing comparison operators, and the ``Widget`` implementor
+    only has to worry about writing ``_cmp_iter``, and making sure the
+    elements in it are also comparable.
+
+    Note: if a class already has ``__eq__``, ``__ne__``, ``__lt__``,
+    ``__le__``, ``__gt__``, ``__ge__``, or ``__hash__`` defined, this
+    decorator will overwrite them.
+
+    If ``set_hash`` is ``False``, this will not overwrite ``__hash__``.
+
+    Raises:
+        TypeError: If the class does not have a ``_cmp_iter`` method
 
-       Raises:
-           TypeError: If the class does not have a ``_cmp_key`` method
     """
-    def setter(name, value):
-        value.__name__ = name
-        setattr(cls, name, value)
-
-    if not has_method(cls, '_cmp_key'):
-        raise TypeError("'%s' doesn't define _cmp_key()." % cls.__name__)
-
-    setter('__eq__',
-           lambda s, o:
-           (s is o) or (o is not None and s._cmp_key() == o._cmp_key()))
-    setter('__lt__',
-           lambda s, o: o is not None and s._cmp_key() < o._cmp_key())
-    setter('__le__',
-           lambda s, o: o is not None and s._cmp_key() <= o._cmp_key())
-
-    setter('__ne__',
-           lambda s, o:
-           (s is not o) and (o is None or s._cmp_key() != o._cmp_key()))
-    setter('__gt__',
-           lambda s, o: o is None or s._cmp_key() > o._cmp_key())
-    setter('__ge__',
-           lambda s, o: o is None or s._cmp_key() >= o._cmp_key())
-
-    setter('__hash__', lambda self: hash(self._cmp_key()))
+    if not has_method(cls, "_cmp_iter"):
```

Comment:
I might add a TODO here to try to check this with a mypy protocol before runtime (https://mypy.readthedocs.io/en/stable/protocols.html).